framework/aster-frame/src/task/atomic.rs

```diff
@@ -0,0 +1,208 @@
+//! Specifically, a kernel panic would be triggered
+//! if such forbidden behaviors are captured from the kernel code in the atomic mode:
+//!
+//! - Any operation that can incur scheduling / context switching:
+//!     - Sleeping;
+//!     - Waiting for I/O;
+//!     - Yielding: yield the processor to other tasks voluntarily;
+//!     - Preemption: get preempted by other tasks;
+//! - And switching to the user space.
+//!
+//! Note that the kernel code is allowed to handle interrupts within the atomic mode.
+//!
+//! The concept of the atomic mode is connected with those of process and interrupt contexts.
+//! A broad classification of execution contexts in a classic monolithic kernel can be:
+//!
+//! | Context Type in Kernel | Process                                 | Interrupt                 |
+//! | ---------------------- | --------------------------------------- | ------------------------- |
+//! | Execution Target       | kernel code on behalf of a user process | interrupt handler         |
+//! | Atomic Requirement     | may switch into or out on demand        | always in the atomic mode |
+//!
+
+use core::{
+    panic,
+    sync::atomic::{AtomicUsize, Ordering::Relaxed},
+};
+extern crate alloc;
+use alloc::{format, string::String};
+
+crate::cpu_local! {
+    static ATOMIC_MODE: AtomicMode = AtomicMode::new(AtomicModeInner::new());
+}
+
+#[derive(Debug, Default)]
+struct AtomicModeInner {
+    /// The number of locks and IRQ-disabled contexts held by the current CPU.
+    cnt: AtomicUsize,
+}
+
+impl AtomicModeInner {
+    const fn new() -> Self {
+        Self {
+            cnt: AtomicUsize::new(0),
+        }
+    }
+    fn inc_cnt(&self) {
+        self.cnt.fetch_add(1, Relaxed);
+    }
+    fn dec_cnt(&self) {
+        self.cnt.fetch_sub(1, Relaxed);
+    }
+    fn cnt(&self) -> usize {
+        self.cnt.load(Relaxed)
+    }
+}
+
+pub struct AtomicMode {
+    inner: AtomicModeInner,
+}
+
+impl AtomicMode {
+    const fn new(inner: AtomicModeInner) -> Self {
+        Self { inner }
+    }
+
+    #[must_use]
+    pub fn new_guard(&self) -> AtomicModeGuard {
+        self.inner.inc_cnt();
+        AtomicModeGuard { private: () }
+    }
+
+    pub fn in_atomic_mode(&self) -> bool {
+        self.inner.cnt() != 0
+    }
+
+    fn on_guard_drop(&self) {
+        self.inner.dec_cnt();
+    }
+
+    pub fn info(&self) -> String {
+        format!("ATOMIC_MODE = {}", self.inner.cnt())
+    }
+}
+
+/// The guard of the atomic mode.
+/// The CPU shall exit the atomic mode only after the last guard object is dropped.
+pub struct AtomicModeGuard {
+    #[allow(dead_code)]
+    // This private field prevents user from constructing values of this type directly.
+    private: (),
+}
+
+impl !Send for AtomicModeGuard {}
+
+impl Drop for AtomicModeGuard {
+    fn drop(&mut self) {
+        ATOMIC_MODE.on_guard_drop();
+    }
+}
+
+/// Enter the atomic mode, and return a guard for exiting the current atomic mode.
```

Comment:
```suggestion
/// Enters the atomic mode, and return a guard for exiting the current atomic mode.
```

Please fix other documentation as well.